### PR TITLE
mrc-363 prototype vuex SPA

### PR DESCRIPTION
**Pull request by hillalex**
We already have a pretty good idea of what `react` involves, so I wanted to try out building a SPA with `vue` to see how it compares. `vuex` is the `redux` equivalent, a flux implementation for `vue`. Also to try out `vue` with `typescript` - for the complexity of a SPA definitely think we should be using types.

It feels pretty simple to grasp given our familiarity with the flux pattern in general, and I still prefer it to `react`, if only because I much prefer writing `html` to `jsx`! What's here took me a few hours (maybe 4) and I need to add tests, but felt pretty easy to get going with so my verdict is positive.

### CODE REVIEW

File: app/static/public/css/style.css [hunk 1]
```diff
@@ -4705,7 +4705,7 @@ button.bg-dark:focus {
 .rounded-lg {
   border-radius: 0.3rem !important; }
 
-.rounded-circle {
+.rounded-circle, .step .btn {
```
[the remark this answers is not in the document]
Comment (hillalex) [reply app/static/public/css/style.css]: It should not be. Correctly spotted.

File: app/static/src/app/components/Stepper.vue [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    import Vue from "vue";
+    import {Dictionary, mapState} from "vuex";
+    import {RootState} from "../types";
+    import Step from "./Step.vue";
+    import Baseline from "./Baseline.vue";
+
+    type StepType = "baseline" | "surveyAndProgram" | "review" | "run" | "results"
+
+    type Status = {
```
[the remark this answers is not in the document]
Comment (hillalex) [reply app/static/src/app/components/Stepper.vue]: Yeah this isn't actually being used here.

File: app/templates/index.ftl [hunk 3]
```diff
@@ -1,43 +1,13 @@
 <!DOCTYPE HTML>
-<html xmlns:th="http://www.thymeleaf.org">
+<html>
```
[the remark this answers is not in the document]
Comment (hillalex) [reply app/templates/index.ftl]: Yeah, at the moment not making much use of Freemarker, I guess these could be vanilla html.

**Comment by hillalex** [comment]
> This seems like a really good approach to me. We'll have decisions to make about how we structure the state for the whole app and per step etc, but the flux pattern seems useful for separating logic in a structured way. I agree that it seems clearer than Redux but maybe that's just because we haven't got into the more complex stuff yet!
> 
> Rob spotted a little issue with the README file: `./app/gradlew :bootRun` should be `./gradlew :bootRun` from the app folder.
> 
> I found that I needed to run webpack manually to get the js to build. But Rob didn't have this issue. (That should just happen as part of bootRun, shouldn't it?) Have a vague memory of something similar happening in the early days of orderlyweb but can't remember what the explanation was!

It won't run automatically by default. In OrderlyWeb I added a task to the gradle build task to compile front-end deps but haven't yet done that here. For now you should run `npm run build` to compile front end, but we should follow the OW pattern and add it to the boot run task!

File: app/static/src/app/store/baseline/baseline.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import { Module } from 'vuex';
```
[the remark this answers is not in the document]
Comment (hillalex) [reply app/static/src/app/store/baseline/baseline.ts]: This seems to be the naming convention based on the tutorials I followed

File: README.md [hunk 5]
```diff
@@ -9,4 +9,5 @@ Requirements:
 1. Clone this repo
 1. Run `npm install` from `app/static`
 1. Run `./scripts/run-dependencies.sh` to start a model API
-1. Run app from your IDE or by `./app/gradlew :bootRun` to serve the app on port 8080
+1. Run `npm run build` from `app/static` to compile front-end dependencies.
+1. Run app from your IDE or by `cd ap && ./gradlew :bootRun` to serve the app on port 8080
```
Comment (EmmaLRussell) [inline README.md]: Should be `cd app..

File: app/static/src/tests/baseline/actions.test.ts [hunk 6]
```diff
@@ -21,18 +21,35 @@ describe("Baseline actions", () => {
...
             expect(commit.mock.calls[0][0]).toStrictEqual({
                 type: "PJNZUploadError",
-                payload: [{code: "e", message: "error message"}]
+                payload: "error message"
+            });
+            done();
+        })
+    });
+
+    it("returns basic error if response is a 40 after failed PJNZ file upload", (done) => {
```
Comment (EmmaLRussell) [inline app/static/src/tests/baseline/actions.test.ts]: A 40? This is testing for a 500.

File: app/static/package.json [hunk 7]
```diff
@@ -26,6 +26,7 @@
     "gulp-clean-css": "^4.0.0",
     "gulp-rename": "^1.4.0",
     "gulp-sass": "^4.0.2",
+    "i": "^0.3.6",
```
Comment (EmmaLRussell) [inline app/static/package.json]: Can this be removed?

**Pull request merged**